### ISSUE TITLE
False positive used-before-assignment for return type

### ISSUE THREAD

**Issue by jp-diegidio** [PyCQA/pylint#4066]
Hello guys, here seems to be an issue with return type annotations conflicting with member names:

### Current behavior

This is a little class to reproduce the problem, the relevant (mis-)behaviour is explained in the comments:

```python
class TestPylint_return_1:

    # Hovering on 'bool' anywhere shows "bool: bool" i.e. the class member!

    def foo(self) -> bool:  # raises 'bool: bool' used-before-assignment!
        return self.bool()

    def bool(self) -> bool:
        return True
```

And this simple work-around does get rid of the warning but shows that the underlying issue remains:

```python
class TestPylintBool_return_2:

    # Hovering on 'bool' anywhere shows "bool: bool"!

    def bool(self) -> bool:
        return True

    def foo(self) -> bool:
        return self.bool()

```

Though of course at that point I have realised that the "underlying issue" may not just be with pylint...

### Expected behavior

Pylint should (at least) recognise that that 'bool' annotating the return type is the 'bool' type and not the 'bool' member of the class, hence not give that warning.

### pylint --version output

pylint 2.6.0
astroid 2.4.2
Python 3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]

Thank you,

Julio

**Comment by Pierre-Sassoulas** [comment]
Hello, thank you for opening an issue.

This is unrelated to the bug you signaled in pylint, but I think your `bool` function could be `__bool__` instead (called `bool(instance)`, instead of `instance.bool()`). In fact it would be the same for `__str__`, and `__int__` versus `str` and `int` the types, so this is a bug that must not happen very often.

**Comment by jp-diegidio** [comment]
> Hello, thank you for opening an issue.
> 
> This is unrelated to the bug you signaled in pylint, but I think your `bool` function could be `__bool__` instead (called `bool(instance)`, instead of `instance.bool()`). In fact it would be the same for `__str__`, and `__int__` versus `str` and `int` the types, so this is a bug that must not happen very often.

Thanks but that was just the proverbial minimal example to reproduce the problem: of course in reality it's not conversions that I am doing, and I don't think a method name that matches a type name is that rare at all, e.g. in a typing validation suite...

Julio
